src/util.c

```diff
@@ -82,14 +85,18 @@ static SECStatus FileToItem(SECItem * dst, PRFileDesc * src)
 	    /* out of memory */
 	    break;
 
-	bytesReadNow = PR_Read(src, dst->data + bytesReadTotal, chunk);
+	/* subtract 1 to leave room for null terminator */
+	bytesReadNow = PR_Read(src, dst->data + bytesReadTotal,
+			       newSize - bytesReadTotal - 1);
+
 	if (bytesReadNow < 0)
 	    /* read error */
 	    break;
 
 	if (bytesReadNow == 0) {
 	    /* EOF */
-	    dst->len = bytesReadTotal;
+	    dst->len = bytesReadTotal + 1;
```

Comment:
Either this should be `dst->len = bytesReadTotal` in case we do not want to include the terminating `NUL` in the binary data, or the next line should be `dst->data[bytesRedaTotal] = '\0'`, so that we write `NUL` right after the read data?

Reply:
Good catch! I think setting `dst->len = bytesReadTotal; dst->data[bytesReadTotal] = '\0';` is probably the way to go.

One question I had is whether `dst->len` should instead contain the length of the buffer allocated as `SECITEM_ReallocItemV2` uses len to determine whether a reallocation is necessary. But in this case I see we just call `free` on the buffer when we are done so it doesn't matter. Should it ultimately be calling `SECITEM_FreeItem` ?

Reply:
Looks like SECITEM_ReallocItemV2 doesn't maintain the actual allocation size.

Reply:
Not that I found it documented anywhere but I believe that `SECItem::len` is the actual size of valid data to be processed.